sdks/go/pkg/beam/core/metrics/store.go

```diff
@@ -145,16 +150,73 @@ type ptCounterSet struct {
 	gauges        map[nameHash]*gauge
 }
 
+// Bundle processing state (START_BUNDLE, PROCESS_BUNDLE, FINISH_BUNDLE)
+type BundleProcState string
+
+const (
+	StartBundle   BundleProcState = "START_BUNDLE"
+	ProcessBundle BundleProcState = "PROCESS_BUNDLE"
+	FinishBundle  BundleProcState = "FINISH_BUNDLE"
+)
+
+// ExecutionState stores the information about a bundle in a particular state.
+type ExecutionState struct {
+	State           BundleProcState
+	IsProcessing    bool // set to true when sent as a response to ProcessBundleProgress Request
+	TotalTimeMillis int64
+}
+
+// ExecutionStateTracker stores information about a bundle for execution time metrics.
+type ExecutionStateTracker struct {
+	CurrentState              BundleProcState
+	State                     ExecutionState
+	NumberOfTransitions       int
+	MillisSinceLastTransition int
+	TransitionsAtLastSample   int
+	LastSampleTimeMillis      int
+}
+
 // Store retains per transform countersets, intended for per bundle use.
 type Store struct {
 	mu  sync.RWMutex
 	css []*ptCounterSet
 
 	store map[Labels]userMetric
+
+	executionStore ExecutionStateTracker
+	stateRegistry  map[Labels][]ExecutionState
 }
 
 func newStore() *Store {
-	return &Store{store: make(map[Labels]userMetric)}
+	return &Store{store: make(map[Labels]userMetric), stateRegistry: make(map[Labels][]ExecutionState)}
+}
+
+// AddState adds an ExecutionState of a PTransform to the stateRegistry.
+func (s *Store) AddState(e ExecutionState, pid string) {
+	label := PTransformLabels(pid)
+	if s.stateRegistry == nil {
+		s.stateRegistry = make(map[Labels][]ExecutionState)
+	}
+	s.stateRegistry[label] = append(s.stateRegistry[label], e)
```

Comment:
AddState is being called in the sample method, and nothing clears these away. That means that each individual sample is retained for the life of the bundle. 
Probably not noticeable in streaming, but in batch, a bundle can last for hours, and that will add up to an astounding number of allocations.

We need to sum up contributions for the same PTransform + State combination (bundle is implied by construction). This can be facilitated by having the state be an int under the hood, and using a fixed sized array as mentioned elsewhere.

To be concrete a Bundle that has 1 DoFn should have at most, 4 exec metrics reported for it (start, process, finish, and total). 4 DoFns would have 16, etc.